### PR TITLE
fix: prevent false label masking when 'assistant' appears in user content

### LINKED ISSUE #2010: Bug in qwen-vl-finetune/qwenvl/data/data_processor.py/preprocess_qwen_visual: Bad things happen when token 77091(assistent) appears in some position
I notice that the `preprocess_qwen_visual` in `data_processor.py` uses token 77091 to define labels:
``` python
    messages = _build_messages(source, base_path)

    full_result = processor.apply_chat_template(
        messages, tokenize=True, return_dict=True, return_tensors="pt"
    )

    input_ids = full_result["input_ids"]
    if isinstance(input_ids, list):
        input_ids = torch.tensor(input_ids).unsqueeze(0)

    labels = torch.full_like(input_ids, IGNORE_INDEX)

    input_ids_flat = input_ids[0].tolist()
    L = len(input_ids_flat)
    pos = 0
    while pos < L:
        if input_ids_flat[pos] == 77091:
            ans_start = pos + 2
            ans_end = ans_start
            while ans_end < L and input_ids_flat[ans_end] != 151645:
                ans_end += 1
            if ans_end < L:
                labels[0, ans_start : ans_end + 2] = input_ids[
                    0, ans_start : ans_end + 2
                ]
                pos = ans_end
        pos += 1

    full_result["labels"] = labels
    full_result["input_ids"] = input_ids
```
However, if token 77091(*assistant* without leading spaces) appears in other position, it will identified as the start of the label and participate in the loss calculation. For example
``` python
from transformers import AutoProcessor
import torch

IGNORE_INDEX = -100

model_path = "Qwen3-VL-8B-Instruct"

processor = AutoProcessor.from_pretrained(model_path)

# print(processor)

message = [
    {"role": "user","content": [{"type": "text", "text": "assistant: false assistant content"},]},
    {"role": "assistant","content": [{"type": "text", "text": "assistant: true assistant content"},],
}]

text = processor.apply_chat_template(
    message, tokenize=False, return_dict=True, return_tensors="pt"
)
print('==================text==================')
print(text)

full_result = processor.apply_chat_template(
    message, tokenize=True, return_dict=True, return_tensors="pt"
)
print('==================ids===================')
print(list(full_result['input_ids'][0].cpu().numpy()))

input_ids = full_result["input_ids"]
if isinstance(input_ids, list):
    input_ids = torch.tensor(input_ids).unsqueeze(0)

labels = torch.full_like(input_ids, IGNORE_INDEX)

input_ids_flat = input_ids[0].tolist()
L = len(input_ids_flat)
pos = 0
while pos < L:
    if input_ids_flat[pos] == 77091:
        ans_start = pos + 2
        ans_end = ans_start
        while ans_end < L and input_ids_flat[ans_end] != 151645:
            ans_end += 1
        if ans_end < L:
            labels[0, ans_start : ans_end + 2] = input_ids[
                0, ans_start : ans_end + 2
            ]
            pos = ans_end
    pos += 1

print('=================labels=================')
print(list(labels[0].cpu().numpy()))
```
will give the results:

Image

Some of the token ids in the user content part is incorrectly exposed to the labels.

### PR DESCRIPTION
**Pull request by Mr-Neutr0n**
## Summary
Fix training data leakage when the word "assistant" appears in user content.

## Problem
In `preprocess_qwen_visual()`, the code uses token 77091 ("assistant" without leading space) to identify the start of labels:

```python
if input_ids_flat[pos] == 77091:  # "assistant"
    ans_start = pos + 2
    ...
```

However, if "assistant" appears anywhere in user content (e.g., "You are an assistant"), it incorrectly triggers label extraction at that position, causing:
- Training data leakage (user content exposed in labels)
- Incorrect loss calculation

**Example:**
```python
message = [
    {"role": "user", "content": "assistant: false assistant content"},
    {"role": "assistant", "content": "true assistant content"},
]
```
The first occurrence of "assistant" in user content would be incorrectly matched.

## Solution
Check for the **full pattern** that marks an assistant turn in the chat template: `<|im_start|>assistant` (token 151644 followed by token 77091), rather than just the word "assistant" alone.

```python
if (
    input_ids_flat[pos] == ASSISTANT_TOKEN
    and pos > 0
    and input_ids_flat[pos - 1] == IM_START_TOKEN
):
```

This correctly identifies only the actual assistant role markers, not the word "assistant" appearing in content.

## Changes
1. Added named constants for special tokens (`IM_START_TOKEN`, `IM_END_TOKEN`, `ASSISTANT_TOKEN`)
2. Updated label extraction to require `<|im_start|>` before `assistant`

Fixes #2010

### CODE REVIEW

**Comment by Mr-Neutr0n** [comment]
Friendly bump! Let me know if there's anything I should update or improve to help move this forward.

**Comment by Mr-Neutr0n** [comment]
Closing this one to tidy up my open pull requests.

It's been open around five months with no review activity, and I'd rather withdraw it than leave it sitting in your queue. Nothing needed from you, and no hard feelings at all.

If the fix is still wanted, this can be reopened, or I'm happy to redo it properly against current main. Apologies for the noise.
